pkg/gcp/gcp.go

```diff
@@ -0,0 +1,178 @@
+		return fmt.Errorf("failed to open submariner internal ports: %v", err)
+	}
+
+	reporter.Succeeded("Submariner internal ports are opened with firewall rule %q on GCP", internalIngress.Name)
+
+	return nil
+}
+
+// CleanupAfterSubmariner clean up submariner cluster environment on GCP
+func (gc *gcpCloud) CleanupAfterSubmariner(reporter api.Reporter) error {
+	ingressName, egressName := generateRuleNames(gc.infraID, publicPortsRuleName)
+	// delete the inbound and outbound firewall rules to close submariner public ports
+	if err := gc.client.DeleteFirewallRule(ingressName); err != nil {
+		return err
+	}
+
+	if err := gc.client.DeleteFirewallRule(egressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner public ports are closed on GCP")
+
+	// delete the inbound and outbound firewall rules to close submariner internal ports
+	internalIngressName, _ := generateRuleNames(gc.infraID, internalPortsRuleName)
+	if err := gc.client.DeleteFirewallRule(internalIngressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner internal ports are closed on GCP")
+
+	return nil
+}
+
+func (gc *gcpCloud) newFirewallRules(name string, ports []api.PortSpec, isInternal bool) (ingress, egress *compute.Firewall) {
+	ingressName, egressName := generateRuleNames(gc.infraID, name)
+
+	allowedPorts := []*compute.FirewallAllowed{}
+	for _, port := range ports {
+		allowedPorts = append(allowedPorts, &compute.FirewallAllowed{
+			IPProtocol: port.Protocol,
+			Ports:      []string{strconv.Itoa(int(port.Port))},
+		})
+	}
+
+	ingress = &compute.Firewall{
+		Name:      ingressName,
+		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
+		Direction: "INGRESS",
+		Allowed:   allowedPorts,
+	}
+
+	// if ports are internal, we make the ports are accessed in internal cluster and egress is not required
+	if isInternal {
+		ingress.TargetTags = []string{
+			fmt.Sprintf("%s-worker", gc.infraID),
+			fmt.Sprintf("%s-master", gc.infraID),
+		}
+		ingress.SourceTags = []string{
+			fmt.Sprintf("%s-worker", gc.infraID),
+			fmt.Sprintf("%s-master", gc.infraID),
+		}
+
+		return ingress, nil
+	}
+
+	return ingress, &compute.Firewall{
+		Name:      egressName,
+		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
+		Direction: "EGRESS",
+		Allowed:   allowedPorts,
+	}
+}
+
+// open expected ports by creating related firewall rule
+// - if the firewall rule is not found, we will create it
+// - if the firewall rule is found and changed, we will update it
+func (gc *gcpCloud) openPorts(rules ...*compute.Firewall) error {
+	for _, rule := range rules {
+		current, err := gc.client.GetFirewallRule(rule.Name)
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 404 {
+			if err := gc.client.InsertFirewallRule(rule); err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if !ruleChanged(current, rule) {
```

Comment:
Please document why is this needed?

IIUC, it seems GCP won't change anything in this case. It seems an unnecessary optimization to me.

Reply:
I will try this on GCP, it GCP won't change anything, I will remove this check

Reply:
test on GCP, remove this check